components/render/atlastexture.cpp

```diff
@@ -0,0 +1,148 @@
+/* Stores many small textures into a large texture (or array of textures)
+ * to allow batched drawing commands that increase performance.
+ * Currently implemented as an array of 2D textures, would be simpler with much
+ * better performance if a proper 2D texture array (GL_TEXTURE_2D_ARRAY) was
+ * used. Author couldn't quite get it to work, parts of the Nuklear GUI would
+ * be missing if the array texture was allocated with > 1 depth (layers)... */
+
+#include "atlastexture.h"
+#include "../../extern/RectangleBinPack/SkylineBinPack.h"
+
+#include <boost/make_unique.hpp>
+#include <misc/assert.h>
+
+namespace Render
+{
+    AtlasTexture::AtlasTexture()
+    {
+        GLint maxTextureSize;
+        GLint maxTextures = 8; // Hardcoded in fragment shader, GL 3.x minimum is 16.
+        glGetIntegerv(GL_MAX_TEXTURE_SIZE, &maxTextureSize);
+
+        mTextureWidth = maxTextureSize;
+        mTextureHeight = maxTextureSize;
+
+        // Limit number of textures to a reasonable level (measured from testing).
+        // Note: Increasing this has a severe impact on performance.
+        GLint estimatedRequiredTextures = (1uLL << 29) / (mTextureWidth * mTextureHeight);
+        mTextureLayers = std::min(estimatedRequiredTextures, maxTextures);
```

Comment:
```suggestion
        mTextureLayers = estimatedRequiredTextures > 0 ? std::min(estimatedRequiredTextures, maxTextures) : maxTextures;
```

Reply:
Doing this leads to a glError (freeablo/components/render/sdl2backend.cpp:694, 0x0506)

Reply:
Try `1` instead of `maxTextures`, e.g:

    mTextureLayers = estimatedRequiredTextures > 0 ? std::min(estimatedRequiredTextures, maxTextures) : 1;

It probably should just round up, e.g:

    // In some helper header
    #define DIVIDE_ROUND_UP(numerator, demoninator) \
            ( ((numerator) + ((demoninator) - 1u)) / (demoninator) )
    
    GLint estimatedRequiredTextures = DIVIDE_ROUND_UP(1uLL << 29, mTextureWidth * mTextureHeight);
C++ probably has build in functionality for this

Reply:
Now it is allocating so much memory it freezes my system (I have 8gb of RAM and a 19gb swap partition).

Reply:
It might be allocating too much vram from your graphics card..
Can you try this instead:

glGetIntegerv(GL_MAX_TEXTURE_SIZE, &maxTextureSize);
maxTextureSize = std::min(maxTextureSize, 16384);
 mTextureWidth = maxTextureSize;

Reply:
Now it worked. Getting 370-400 fps. Wow.